docs/source-2.0/additional-specs/rules-engine/specification.rst

```diff
@@ -74,6 +74,184 @@ or :ref:`error rules, <rules-engine-endpoint-rule-set-error-rule>` with an
 empty set of conditions to provide a more meaningful default or error depending
 on the scenario.
 
+.. smithy-trait:: smithy.rules#endpointBdd
+.. _smithy.rules#endpointBdd-trait:
+
+----------------------------------
+``smithy.rules#endpointBdd`` trait
+----------------------------------
+
+.. warning:: Experimental
+
+    This trait is experimental and subject to change.
+
+Summary
+    Defines a `Binary Decision Diagram (BDD) <https://en.wikipedia.org/wiki/Binary_decision_diagram>`_ representation
+    of endpoint rules for efficient runtime evaluation.
+Trait selector
+    ``service``
+Value type
+    ``structure``
+
+The ``endpointBdd`` trait provides a BDD representation of endpoint rules, optimizing runtime evaluation by
+eliminating redundant condition evaluations and reducing the decision tree to a minimal directed acyclic graph.
+This trait is an alternative to ``endpointRuleSet`` that trades compile-time complexity for significantly improved
+runtime performance and reduced artifact sizes.
+
+.. note::
+
+   The ``endpointBdd`` trait can be generated from an ``endpointRuleSet`` trait through compilation. Services may
```

Comment:
Since we can go from rules -> bdd, but (as far as I know) not the other way from bdd -> rules, I might make an argument that we (for now at least) prefer endpointRules trait since consumers can use a simple projection to remove endpoint rules or compile them to bdd.

Reply:
We should eventually stop shipping the tree in the models because they cause cold start issues with SDKs that use models dynamically (e.g., botocore). That's what the guidance is steering us toward here.

As for coverage, we run the same modeled test cases against the BDD too. We don't do coverage analysis there, but I can see if that's possible.

Reply:
I just pushed up a BDD coverage checker, similar to the ruleset checker. When run with S3 and its test cases, we get 100% coverage (yay)